contributors/design-proposals/cloud-provider-refactoring.md

```diff
@@ -0,0 +1,168 @@
+
+This approach requires kubelet to be started with known taints. This will make the node unschedulable until these taints are removed. The external cloud controller manager will asynchronously update the node objects and remove the taints. 
+
+### 4. Strategy for refactoring Kube-ApiServer
+
+Kube-apiserver uses the cloud provider for two purposes
+
+1. Distribute SSH Keys - This can be moved to the cloud dependent controller manager
+2. Admission Controller for PV - This can be refactored using the taints approach used in Kubelet
+
+### 5. Strategy for refactoring Volumes
+
+Volumes need cloud providers, but they only need SPECIFIC cloud providers.  The real problem is that cloud providers can take params from cloud config. The majority of the logic for volume management, currently resides in the controller-manager.
+
+There is an undergoing effort to move all of the volume logic from the controller-manager into plugins called Flex Volumes. In the Flex volumes world, all of the vendor specific code will be packaged in a separate binary as a plugin. After discussing with @thockin, this was decidedly the best approach to remove all cloud provider dependency for volumes out of kubernetes core.
+
+### 6. Deployment, Upgrades and Downgrades
+
+This change will introduce new binaries to the list of binaries required to run kubernetes. The change will be designed such that these binaries can be installed via `kubectl apply -f` and the appropriate instances of the binaries will be running. 
+
+##### 6.1 Upgrading kubelet and proxy
+
+The kubelet and proxy runs on every node in the kubernetes cluster. Based on your setup (systemd/other), you can follow the normal upgrade steps for it. This change does not affect the kubelet and proxy upgrade steps for your setup.
+
+##### 6.2 Upgrading plugins
+
+Plugins such as cni, flex volumes can be upgraded just as you normally upgrade them. This change does not affect the plugin upgrade steps for your setup.
+
+###### 6.3 Upgrading kubernetes core services
+
+The master node components (kube-controller-manager,kube-scheduler, kube-apiserver etc.) can be upgraded just as you normally upgrade them. This change does not affect the plugin upgrade steps for your setup.
+
+##### 6.4 Applying the cloud-controller-manager
+
+This is the only step that is different in the upgrade process. In order to complete the upgrade process, you need to apply the cloud-controller-manager deployment to the setup. A deployment descriptor file will be provided with this change. You need to apply this change using
+
+```
+kubectl apply -f cloud-controller-manager.yml
```

Comment:
What happens when we update the master and get the new controller, but the nodes have not been updated?  How does the taints/initializers stuff work in the interim?  Do we need an interlock like we have done before?  e.g. one release has both old and new master logic and switches only when all nodes report updated kubelets, and then the subsequent release removes old logic?